codec/src/codec.rs

```diff
@@ -16,6 +16,25 @@ pub trait RangeConfig: Config + RangeBounds<usize> {}
 // Automatically implement `RangeConfig` for matching types.
 impl<T: Config + RangeBounds<usize>> RangeConfig for T {}
 
+/// Trait for types with a fixed encoded size.
+pub trait FixedSize {
+    /// The size of the encoded value (in bytes).
+    const SIZE: usize;
+}
+
+/// Trait for types that may have a variable or fixed encoded size.
+pub trait EncodeSize {
+    /// Returns the encoded size of this value (in bytes).
+    fn encode_size(&self) -> usize;
+}
+
+// Automatically implement `EncodeSize` for types with a fixed size.
+impl<T: FixedSize> EncodeSize for T {
+    fn encode_size(&self) -> usize {
+        Self::SIZE
+    }
+}
+
```

Comment:
This should hopefully make it a lot clearer what's going on and make it more obvious that clients have to implement one or the other, but not both